File: backend/app/services/product_search.py
from datetime import date

from app.models.schemas import BuySuggestion, ProductListing


def _safe_float(value) -> float | None:
    try:
        return float(value) if value is not None else None
    except (ValueError, TypeError):
        return None


def _safe_int(value) -> int | None:
    try:
        return int(value) if value is not None else None
    except (ValueError, TypeError):
        return None


def _fmt_weight(value) -> str | None:
    """Weight__c may be numeric (e.g. 0.5) or text (e.g. "500 g"). Render it as a
    clean string for display, dropping a trailing ".0" on whole numbers."""
    if value is None:
        return None
    if isinstance(value, (int, float)):
        f = float(value)
        return str(int(f)) if f.is_integer() else str(f)
    text = str(value).strip()
    return text or None


def _ci_get(record: dict, key: str):
    """Case-insensitive lookup. Salesforce returns custom field names with
    whatever case is stored in metadata — defending against both `Title__c`
    and `title__c` keeps us decoupled from org-specific naming."""
    if key in record:
        return record[key]
    target = key.lower()
    for k, v in record.items():
        if k.lower() == target:
            return v
    return None


def _parse_sf_date(value) -> date | None:
    if value is None:
        return None
    if isinstance(value, date):
        return value
    try:
        return date.fromisoformat(str(value)[:10])
    except (ValueError, TypeError):
        return None


_RESTOCK_THRESHOLD_DAYS = 7
_FREQUENT_THRESHOLD = 3
_MISSING_RANK_SCORE = -(10**9)


def _derive_suggestion(
    times: int | None,
    last_ordered: date | None,
    today: date,
) -> tuple[BuySuggestion | None, str | None]:
    if times is None or times <= 0:
        return "new", "Never ordered before"

    if times >= _FREQUENT_THRESHOLD:
        if last_ordered is not None:
            days = (today - last_ordered).days
            return "frequent", f"Bought {times}x, last {days} days ago"
        return "frequent", f"Bought {times}x"

    if last_ordered is None:
        return "restock", f"Bought {times}x, last order date unknown"

    days = (today - last_ordered).days
    if days >= _RESTOCK_THRESHOLD_DAYS:
        return "restock", f"Bought {times}x, last {days} days ago"
    return "recent", f"Bought {times}x, last {days} days ago"


def _score(record: dict) -> tuple[int, int]:
    """Determines order within a source group. The first item per group is shown
    as 'Top match' in the UI.

    Primary: number of times the user has purchased this product (desc).
    Tie-break: vendor rank__c (asc — #1 wins over #7). Records with no rank
    sort last among ties; records with no purchase history all tie at 0.
    """
    times = _safe_int(_ci_get(record, "Number_Of_Times_Purchased__c")) or 0
    rank_value = _safe_int(_ci_get(record, "Rank__c"))
    rank_score = -rank_value if rank_value is not None else _MISSING_RANK_SCORE
    return (times, rank_score)


def _normalize(record: dict, today: date | None = None) -> ProductListing:
    current_price = _safe_float(_ci_get(record, "Current_Price__c"))
    original_price = _safe_float(_ci_get(record, "Original_Price__c"))
    last_purchased_price = _safe_float(_ci_get(record, "Last_Purchased_Price__c"))
    discount = _safe_int(_ci_get(record, "Discount__c"))

    if (
        discount is None
        and current_price is not None
        and original_price is not None
        and original_price > 0
        and original_price > current_price
    ):
        discount = round((1 - current_price / original_price) * 100)

    rating_value = _ci_get(record, "Rating__c")

    times_purchased = _safe_int(_ci_get(record, "Number_Of_Times_Purchased__c"))
    last_ordered_raw = _ci_get(record, "Last_Ordered_Date__c")
    last_ordered_date = _parse_sf_date(last_ordered_raw)
    buy_suggestion, suggestion_reason = _derive_suggestion(
        times_purchased,
        last_ordered_date,
        today or date.today(),
    )

    return ProductListing(
        id=_ci_get(record, "Id") or "",
        title=_ci_get(record, "Title__c") or _ci_get(record, "Name") or "",
        source=_ci_get(record, "Source__c") or "",
        current_price=current_price,
        original_price=original_price,
        last_purchased_price=last_purchased_price,
        discount=discount,
        rating=str(rating_value) if rating_value is not None else None,
        review_count=_safe_int(_ci_get(record, "Review_Count__c")),
        rank=_safe_int(_ci_get(record, "Rank__c")),
        product_url=_ci_get(record, "Product_URL__c"),
        image_url=_ci_get(record, "Image_URL__c"),
        availability=_ci_get(record, "Availability__c"),
        weight=_fmt_weight(_ci_get(record, "Weight__c")),
        last_ordered_date=str(last_ordered_raw)[:10] if last_ordered_raw else None,
        times_purchased=times_purchased,
        buy_suggestion=buy_suggestion,
        suggestion_reason=suggestion_reason,
    )


def rank_and_group(
    records: list[dict],
    query: str,
    per_source: int = 3,
) -> list[ProductListing]:
    from app.services.salesforce import _filter_tokens
    query_tokens = _filter_tokens(query.split()) if query else []
    if not query_tokens:
        query_tokens = query.lower().split() if query else []
    query_tokens = [t.lower() for t in query_tokens]

    def get_relevance(record: dict) -> int:
        title = _ci_get(record, "Title__c") or _ci_get(record, "Name") or ""
        title_lower = title.lower()
        return sum(1 for t in query_tokens if t in title_lower)

    groups: dict[str, list[tuple]] = {}
    has_any_match = any(get_relevance(r) > 0 for r in records) if query_tokens else False

    for record in records:
        source = _ci_get(record, "Source__c") or "Unknown"
        relevance = get_relevance(record)
        # Discard completely irrelevant results that matched a loose OR query
        # only if we found at least one record that actually matches the keywords
        if has_any_match and relevance == 0:
            continue
        groups.setdefault(source, []).append(((relevance, _score(record)), record))

    result: list[ProductListing] = []
    for _source, items in groups.items():
        # Sort by relevance desc, then times desc, then rank_score desc (asc rank)
        items.sort(key=lambda x: (x[0][0], x[0][1][0], x[0][1][1]), reverse=True)
        for _, record in items[:per_source]:
            result.append(_normalize(record))
    return result

File: worker/src/lib/product_search.ts
import type { BuySuggestion, ProductListing } from "../models/schemas";
import { filterTokens } from "./salesforce";

export type SfRecord = Record<string, unknown>;

const RESTOCK_THRESHOLD_DAYS = 7;
const FREQUENT_THRESHOLD = 3;
const MISSING_RANK_SCORE = -(10 ** 9);

export function safeFloat(value: unknown): number | null {
  if (value === null || value === undefined || value === "") return null;
  const n = Number(value);
  return Number.isFinite(n) ? n : null;
}

export function safeInt(value: unknown): number | null {
  const f = safeFloat(value);
  return f === null ? null : Math.trunc(f);
}

function fmtWeight(value: unknown): string | null {
  if (value === null || value === undefined) return null;
  if (typeof value === "number") {
    return Number.isInteger(value) ? String(value) : String(value);
  }
  const text = String(value).trim();
  return text || null;
}

/** Case-insensitive field lookup (Salesforce custom-field casing varies by org). */
export function ciGet(record: SfRecord, key: string): unknown {
  if (key in record) return record[key];
  const target = key.toLowerCase();
  for (const k of Object.keys(record)) {
    if (k.toLowerCase() === target) return record[k];
  }
  return undefined;
}

function utcDateOnly(d: Date): number {
  return Date.UTC(d.getUTCFullYear(), d.getUTCMonth(), d.getUTCDate());
}

export function utcToday(): number {
  return utcDateOnly(new Date());
}

/** Parse a Salesforce date (YYYY-MM-DD…) to a UTC-midnight epoch, or null. */
export function parseSfDate(value: unknown): number | null {
  if (value === null || value === undefined) return null;
  const s = String(value).slice(0, 10);
  const m = /^(\d{4})-(\d{2})-(\d{2})$/.exec(s);
  if (!m) return null;
  return Date.UTC(Number(m[1]), Number(m[2]) - 1, Number(m[3]));
}

export function deriveSuggestion(
  times: number | null,
  lastOrdered: number | null,
  today: number,
): [BuySuggestion | null, string | null] {
  if (times === null || times <= 0) return ["new", "Never ordered before"];

  if (times >= FREQUENT_THRESHOLD) {
    if (lastOrdered !== null) {
      const days = Math.floor((today - lastOrdered) / 86400_000);
      return ["frequent", `Bought ${times}x, last ${days} days ago`];
    }
    return ["frequent", `Bought ${times}x`];
  }

  if (lastOrdered === null) {
    return ["restock", `Bought ${times}x, last order date unknown`];
  }

  const days = Math.floor((today - lastOrdered) / 86400_000);
  if (days >= RESTOCK_THRESHOLD_DAYS) {
    return ["restock", `Bought ${times}x, last ${days} days ago`];
  }
  return ["recent", `Bought ${times}x, last ${days} days ago`];
}

/** Order within a source group: times purchased desc, then vendor rank asc. */
function scoreRecord(record: SfRecord): [number, number] {
  const times = safeInt(ciGet(record, "Number_Of_Times_Purchased__c")) ?? 0;
  const rankValue = safeInt(ciGet(record, "Rank__c"));
  const rankScore = rankValue !== null ? -rankValue : MISSING_RANK_SCORE;
  return [times, rankScore];
}

export function parseRating(value: unknown): string | null {
  if (value === null || value === undefined) return null;
  if (typeof value === "number") return String(value);
  if (typeof value === "object") {
    const o = value as Record<string, unknown>;
    const val = o.value ?? o.shortDisplayString ?? o.displayString;
    return val !== null && val !== undefined ? String(val) : null;
  }
  const valStr = String(value).trim();
  if (valStr.startsWith("{") && valStr.endsWith("}")) {
    try {
      const cleaned = valStr
        .replace(/'/g, '"')
        .replace(/True/g, "true")
        .replace(/False/g, "false")
        .replace(/None/g, "null");
      const parsed = JSON.parse(cleaned) as Record<string, unknown>;
      const val = parsed.value ?? parsed.shortDisplayString ?? parsed.displayString;
      if (val !== null && val !== undefined) return String(val);
    } catch {
      // fall through to the raw string
    }
  }
  return valStr;
}

export function normalize(record: SfRecord, today?: number): ProductListing {
  const currentPrice = safeFloat(ciGet(record, "Current_Price__c"));
  const originalPrice = safeFloat(ciGet(record, "Original_Price__c"));
  const lastPurchasedPrice = safeFloat(ciGet(record, "Last_Purchased_Price__c"));
  let discount = safeInt(ciGet(record, "Discount__c"));

  if (
    discount === null &&
    currentPrice !== null &&
    originalPrice !== null &&
    originalPrice > 0 &&
    originalPrice > currentPrice
  ) {
    discount = Math.round((1 - currentPrice / originalPrice) * 100);
  }

  const timesPurchased = safeInt(ciGet(record, "Number_Of_Times_Purchased__c"));
  const lastOrderedRaw = ciGet(record, "Last_Ordered_Date__c");
  const lastOrderedDate = parseSfDate(lastOrderedRaw);
  const [buySuggestion, suggestionReason] = deriveSuggestion(
    timesPurchased,
    lastOrderedDate,
    today ?? utcDateOnly(new Date()),
  );

  return {
    id: (ciGet(record, "Id") as string) || "",
    title: ((ciGet(record, "Title__c") ?? ciGet(record, "Name")) as string) || "",
    source: (ciGet(record, "Source__c") as string) || "",
    origin: "catalog",
    current_price: currentPrice,
    original_price: originalPrice,
    last_purchased_price: lastPurchasedPrice,
    discount,
    rating: parseRating(ciGet(record, "Rating__c")),
    review_count: safeInt(ciGet(record, "Review_Count__c")),
    rank: safeInt(ciGet(record, "Rank__c")),
    product_url: (ciGet(record, "Product_URL__c") as string) ?? null,
    image_url: (ciGet(record, "Image_URL__c") as string) ?? null,
    availability: (ciGet(record, "Availability__c") as string) ?? null,
    weight: fmtWeight(ciGet(record, "Weight__c")),
    last_ordered_date: lastOrderedRaw ? String(lastOrderedRaw).slice(0, 10) : null,
    times_purchased: timesPurchased,
    buy_suggestion: buySuggestion,
    suggestion_reason: suggestionReason,
  };
}

/** Tokenize a query the same way the SOQL builder does: split on whitespace,
 * drop stopwords, lowercase. Falls back to the raw split if every token is a
 * stopword (so a query of only filler still searches for something). */
export function queryTokens(query: string): string[] {
  let toks = query ? filterTokens(query.split(/\s+/)) : [];
  if (toks.length === 0 && query) toks = query.split(/\s+/);
  return toks.map((t) => t.toLowerCase());
}

/** How many of the query tokens appear (as substrings) in a title. */
export function relevanceOfTitle(title: string, tokens: string[]): number {
  const t = title.toLowerCase();
  return tokens.reduce((n, tok) => (t.includes(tok) ? n + 1 : n), 0);
}

/** Minimum relevance a row must clear to count as on-topic, given the observed
 * relevances of the candidate set. If ANY row matches every token, require a
 * full match — this drops brand-only/type-only partials (e.g. "Nandini Curd"
 * for "nandini butter" once a real "Nandini Butter" is present). Otherwise
 * require ≥1 token; if nothing matches at all, keep everything. */
export function minRelevance(relevances: number[], nTokens: number): number {
  if (nTokens === 0) return 0;
  if (relevances.some((r) => r === nTokens)) return nTokens;
  if (relevances.some((r) => r > 0)) return 1;
  return 0;
}

/** Drop already-normalized listings that aren't relevant to the query, using the
 * same rule as the catalog ranking. Used for live scraper rows, which the store's
 * own fuzzy search otherwise returns loosely related to the query. */
export function filterRelevant(items: ProductListing[], query: string): ProductListing[] {
  const tokens = queryTokens(query);
  if (tokens.length === 0 || items.length === 0) return items;
  const rels = items.map((p) => relevanceOfTitle(p.title, tokens));
  const min = minRelevance(rels, tokens.length);
  return items.filter((_, i) => rels[i] >= min);
}

/** Score, group by source, keep top `perSource`. Mirrors product_search.rank_and_group:
 * relevance-first (require a full-token match when one exists, else drop
 * zero-relevance rows), then times, then rank. */
export function rankAndGroup(records: SfRecord[], query: string, perSource = 3): ProductListing[] {
  const tokens = queryTokens(query);

  const relevanceOf = (record: SfRecord): number => {
    const title = String((ciGet(record, "Title__c") ?? ciGet(record, "Name")) ?? "");
    return relevanceOfTitle(title, tokens);
  };

  const relevances = records.map(relevanceOf);
  const min = minRelevance(relevances, tokens.length);

  type Scored = { relevance: number; times: number; rankScore: number; record: SfRecord };
  const groups = new Map<string, Scored[]>();
  records.forEach((record, i) => {
    if (relevances[i] < min) return;
    const source = (ciGet(record, "Source__c") as string) || "Unknown";
    const [times, rankScore] = scoreRecord(record);
    const bucket = groups.get(source) ?? [];
    bucket.push({ relevance: relevances[i], times, rankScore, record });
    groups.set(source, bucket);
  });

  const result: ProductListing[] = [];
  for (const items of groups.values()) {
    items.sort((a, b) => b.relevance - a.relevance || b.times - a.times || b.rankScore - a.rankScore);
    for (const item of items.slice(0, perSource)) result.push(normalize(item.record));
  }
  return result;
}
